src/comp_utils.c

```diff
@@ -1,51 +1,176 @@
+static size_t mem_left = 0;
+static void *heap_header = NULL;
+static void *heap_start = NULL;
+const static size_t block_metadata_sz = sizeof(void *) + sizeof(size_t);
 
 #define NON_COMP_DEFAULT_SIZE (1024 * 1024 * 1024) // 1 GB
 
+/*******************************************************************************
+ * Helper functions
+ ******************************************************************************/
+
+static inline void *
+get_next(void *addr)
+{
+    return *(void **) ((char *) addr - sizeof(void *));
+}
+
+static inline size_t
+get_size(void *addr)
+{
+    return *(size_t *) ((char *) addr - block_metadata_sz);
+}
+
+static inline size_t
+get_next_slot_size(void *addr)
+{
+    return (size_t) get_next(addr) - (uintptr_t) addr - get_size(addr);
+}
+
+static inline void
+set_next(void *addr, void *next)
+{
+    memcpy((char *) addr - sizeof(void *), &next, sizeof(void *));
+}
+
+static inline void
+set_size(void *addr, size_t size)
+{
+    memcpy((char *) addr - block_metadata_sz, &size, sizeof(size_t));
+}
+
+static inline void
+make_new_metadata(void *addr, size_t size, void *next, void *prev)
+{
+    set_next(addr, next);
+    set_size(addr, size);
+    if (prev)
+    {
+        set_next(prev, addr);
+    }
+}
+
+static inline void
+clear_block(void *addr)
+{
+    memset((char *) addr - block_metadata_sz, 0, get_size(addr));
+}
+
+/*******************************************************************************
+ * Main functions
+ ******************************************************************************/
+
 void *
 malloc(size_t to_alloc)
 {
-    if (!malloc_ptr)
+    if (!heap_header)
     {
         void *__capability ddc = cheri_ddc_get();
+        void *mem_begin;
         if (cheri_base_get(ddc) == 0)
         {
-            malloc_ptr = mmap(0, NON_COMP_DEFAULT_SIZE, PROT_WRITE | PROT_READ,
+            mem_begin = mmap(0, NON_COMP_DEFAULT_SIZE, PROT_WRITE | PROT_READ,
                 MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-            heap_mem_left = NON_COMP_DEFAULT_SIZE;
-            if (malloc_ptr == MAP_FAILED)
+            if (mem_begin == MAP_FAILED)
             {
-                err(1, "Failed `mmap`");
+                err(1, "comp_utils: Failed `mmap`");
             }
+            mem_left = NON_COMP_DEFAULT_SIZE;
         }
         else
         {
-            malloc_ptr = (char *) cheri_address_get(ddc);
+            mem_begin = (char *) cheri_address_get(ddc);
             // TODO move heap to the end of the compartment; currently, it's at
             // the beginning of the memory scratch area
-            heap_mem_left = cheri_length_get(ddc) - cheri_offset_get(ddc);
+            mem_left = cheri_length_get(ddc) - cheri_offset_get(ddc);
         }
+        heap_start = (char *) mem_begin + sizeof(size_t) + sizeof(void *);
+        heap_header = heap_start;
     }
 
-    if (to_alloc > heap_mem_left)
+    /* We put some metadata at the beginning of each allocated block
+    // * size_t size
```

Comment:
Do we ever need `size`? Can we derive it from `next_block - current_block` or ... ?

Reply:
If we're `free`ing the last block in the chain, then we wouldn't be able to figure out its size.

Reply:
OK, it's not a big deal for now, but it's definitely something where we can learn from existing `malloc`s as to how to optimise this.